WIP: Z: Add array cmp short circuit

ehsankianifar opened the pull request:
@r30shah

Spencer-Comin commented:
@ehsankianifar could you add a description of this change?

r30shah commented:
@ehsankianifar Please paste the tracelog snippet from the Instruction Selection.

r30shah commented:
Also please see https://github.com/eclipse/omr/blob/master/CONTRIBUTING.md for commit guidelines.

As Spencer already mentioned, please add description. In addition to that, a commit should contain clear header and body depicting the intentions behind the change.

ehsankianifar commented:
**What has changed:**
arraycmp and arraycmplen follows different paths to generate result. some paths call helpers which are shared by other methods like memcmp!
1. arraycmplenEvaluator has the implementation for arraycmplen when SIMD is not used. I just compared the address registers and jump to the end if addresses are equal. The instructions looks like this after my changes:
```
 [     0x2aa3e5cb130]	                       proc    	
 [     0x2aa3e5cb230]	                       fence   Relative [ 0x2aa3e55b9b0 ] BBStart <block_2> (frequency 10000)	
 [     0x2aa3e5cb830]	                       LG      &GPR_0017, Parm[Parm  0<parm 0 L>] ?+0(GPR15)
 [     0x2aa3e5e97e0]	                       LG      &GPR_0019, Parm[Parm  1<parm 1 L>] ?+8(GPR15)
 [     0x2aa3e5e9aa0]	                       LG      GPR_0023, Parm[Parm  2<parm 2 J>] ?+16(GPR15)
 [     0x2aa3e5ea0a0]	                       LGR     GPR_0026,GPR_0023	
 [     0x2aa3e5ea150]	                       CLR     &GPR_0017,&GPR_0019	
 [     0x2aa3e5ea200]	                       BRC     BH(0x8), Label L0016	
 [     0x2aa3e5ea2c0]	                       LGR     GPR_0024,GPR_0023	
 [     0x2aa3e5ea370]	                       LGR     GPR_0025,GPR_0023	
 [     0x2aa3e5ea420]	                       CLCL    &GPR_0017,&GPR_0019	
 [     0x2aa3e5eaa60]	                       assocreg
 [     0x2aa3e5ea4d0]	                       SGR     GPR_0026,GPR_0024	
 POST:
 {LegalEvenOfPair:&GPR_0017:R} {LegalOddOfPair:GPR_0024:R} {EvenOddPair:&GPR_0017:GPR_0024:R} {LegalEvenOfPair:&GPR_0019:R} {LegalOddOfPair:GPR_0025:R} {EvenOddPair:&GPR_0019:GPR_0025:R}
 [     0x2aa3e5eab10]	                       Label L0016:	
 [     0x2aa3e5eb230]	                       assocreg
 [     0x2aa3e5eac80]	                       retn    	
 POST:
 {GPR2:GPR_0026:R}
 [     0x2aa3e5eb9a0]	                       fence   Relative [ 0x2aa3e55b9b4 ] BBEnd </block_2>	
 [     0x2aa3e5ebf50]	                       assocreg
```

ehsankianifar commented:
2. arraycmp unsigned calls arraycmpHelper method. The arraycmpHelper follows 3 diffrent paths: 
2.1. if the address nodes are equal or the length is constant and zero, it jumps to the conclusion! I did not change this path since it is already the shortest!
2.2. if load is constant and less than 4kB, it uses a set of clc instructions based on the compare length!
2.3. finally if the length is variable or the length is greater than 4k, it used a different set of instructions based on flags and other conditions.
I extracted tha case 2.2 and 2.3 to a big else condition and added compare and jump instruction on the top and the label on the bottom! since the helper method was used by other evaluators, it may or may not need a result register and may request to jump to a provided label! I set the result register if it is required and jump either to the end of the instructions or to the provided label in case of equal addresses. If the addresses are not equal, it follows the old path!

ehsankianifar commented:
The instructions for the case 2.2 mentioned above looks like this after my changes:
note the first CLR and BRC instruction! GPR_0017 is the result register!
```
 [     0x2aa28a6dfb0]	                       proc    	
 [     0x2aa28a6e0b0]	                       fence   Relative [ 0x2aa289fe9b0 ] BBStart <block_2> (frequency 10000)	
 [     0x2aa28a6e520]	                       XGR     GPR_0017,GPR_0017	
 [     0x2aa28a8c880]	                       LG      &GPR_0018, Parm[Parm  0<parm 0 L>] ?+0(GPR15)
 [     0x2aa28a8cad0]	                       LG      &GPR_0020, Parm[Parm  1<parm 1 L>] ?+8(GPR15)
 [     0x2aa28a8cb80]	                       CLR     &GPR_0018,&GPR_0020	
 [     0x2aa28a8cc30]	                       BRC     BH(0x8), Label L0016	
 [     0x2aa28a8d110]	                       CLC      Method[file:line:name] 0(200,&GPR_0018), Method[file:line:name] 0(&GPR_0020)
 [     0x2aa28a8d2f0]	                       Label L0019:	# (Start of internal control flow)	
 [     0x2aa28a8d1d0]	                       BRC     MASK9(0x8), Label L0018	
 [     0x2aa28a8d950]	                       assocreg
 [     0x2aa28a8d3b0]	                       Label L0017:	
 POST:
 {AssignAny:GPR_0017:R}
 [     0x2aa28a8da00]	                       LGHI    GPR_0017,0x2	
 [     0x2aa28a8dac0]	                       BRC     MASK3(0x2), Label L0018	
 [     0x2aa28a8db80]	                       LGHI    GPR_0017,0x1	
 [     0x2aa28a8e1e0]	                       assocreg
 [     0x2aa28a8dc40]	                       Label L0018:	# (End of internal control flow)	
 POST:
 {AssignAny:GPR_0017:R}
 [     0x2aa28a8e290]	                       Label L0016:	
 [     0x2aa28a8e400]	                       LGFR    GPR_0017,GPR_0017	
 [     0x2aa28a8ea60]	                       assocreg
 [     0x2aa28a8e4b0]	                       retn    	
 POST:
 {GPR2:GPR_0017:R}
 [     0x2aa28aaccc0]	                       fence   Relative [ 0x2aa289fe9b4 ] BBEnd </block_2>	
 [     0x2aa28aad270]	                       assocreg
```

ehsankianifar commented:
The instructions for the case 2.3 mentioned above looks like this:
```
 [     0x2aa2315a130]	                       proc    	
 [     0x2aa2315a230]	                       fence   Relative [ 0x2aa230ea9b0 ] BBStart <block_2> (frequency 10000)	
 [     0x2aa2315a700]	                       XGR     GPR_0017,GPR_0017	
 [     0x2aa23178880]	                       LG      &GPR_0018, Parm[Parm  0<parm 0 L>] ?+0(GPR15)
 [     0x2aa23178ad0]	                       LG      &GPR_0020, Parm[Parm  1<parm 1 L>] ?+8(GPR15)
 [     0x2aa23178b80]	                       CLR     &GPR_0018,&GPR_0020	
 [     0x2aa23178c30]	                       BRC     BH(0x8), Label L0016	
 [     0x2aa23178fe0]	                       L       GPR_0024, Parm[Parm  2<parm 2 I>] ?+16(GPR15)
 [     0x2aa23179220]	                       AHI     GPR_0024,0xffff	
 [     0x2aa23179400]	                       Label L0018:	# (Start of internal control flow)	
 [     0x2aa231792e0]	                       BRC     BM(0x4), Label L0017	
 [     0x2aa231795e0]	                       LR      GPR_0025,GPR_0024	
 [     0x2aa23179690]	                       SRA     GPR_0025,8
 [     0x2aa23179750]	                       BRC     BH(0x8), Label L0020	
 [     0x2aa23179990]	                       Label L0019:	
 [     0x2aa23179a50]	                       CLC     #389 0(256,&GPR_0018),#390 0(&GPR_0020)
 [     0x2aa23179b10]	                       BRC     MASK7(0x6), Label L0021	
 [     0x2aa23179c90]	                       LA      &GPR_0018,#391 256(&GPR_0018)
 [     0x2aa23179e00]	                       LA      &GPR_0020,#392 256(&GPR_0020)
 [     0x2aa23179eb0]	                       BRCT    GPR_0025,Label L0019	
 [     0x2aa23179f70]	                       Label L0020:	
 [     0x2aa23198710]	                       LARL    GPR_0026, &<LiteralPool Base Address>	 ; LoadLitPool
 [     0x2aa231b8e30]	                       assocreg
 [     0x2aa231b88a0]	                       EX      GPR_0024,#395 0(GPR_0026)
 POST:
 {AssignAny:GPR_0017:R} {AssignAny:&GPR_0018:R} {AssignAny:&GPR_0020:R} {AssignAny:GPR_0024:R} {AssignAny:GPR_0025:R} {AssignAny:GPR_0026:R}
 [     0x2aa231b8ee0]	                       Label L0021:	
 [     0x2aa231b8fa0]	                       BRC     MASK9(0x8), Label L0017	
 [     0x2aa231b9060]	                       LGHI    GPR_0017,0x2	
 [     0x2aa231b9120]	                       BRC     MASK3(0x2), Label L0017	
 [     0x2aa231b91e0]	                       LGHI    GPR_0017,0x1	
 [     0x2aa231b9960]	                       assocreg
 [     0x2aa231b93c0]	                       Label L0017:	# (End of internal control flow)	
 POST:
 {AssignAny:GPR_0017:R} {AssignAny:&GPR_0018:R} {AssignAny:&GPR_0020:R} {AssignAny:GPR_0024:R} {AssignAny:GPR_0025:R} {AssignAny:GPR_0026:R}
 [     0x2aa231b9a10]	                       Label L0016:	
 [     0x2aa231b9b80]	                       LGFR    GPR_0017,GPR_0017	
 [     0x2aa231ba1e0]	                       assocreg
 [     0x2aa231b9c30]	                       retn    	
 POST:
 {GPR2:GPR_0017:R}
 [     0x2aa231bad40]	                       fence   Relative [ 0x2aa230ea9b4 ] BBEnd </block_2>	
 [     0x2aa231bb2f0]	                       assocreg
```

ehsankianifar commented:
3. when the sign flag is set (-1/0/1 result is expected) the arraycmpEvaluator uses either MemCmpConstLenSignMacroOp or MemCmpVarLenSignMacroOp classes depending on the length being constant or variable! Those classes are also used by other evaluators but in all cases it must be acceptable to short circuit if the addresses are equal. The OpMemToMem.cpp implements the `generate` method of those classes which is responsible for generating instructions. Like other code paths, I added a compare and branch instructions on top to jump if addresses are equal.
The instructions looks like this for const load:
```
 [     0x2aa13a95fb0]	                       proc    	
 [     0x2aa13a960b0]	                       fence   Relative [ 0x2aa13a269b0 ] BBStart <block_2> (frequency 10000)	
 [     0x2aa13a967d0]	                       LG      &GPR_0017, Parm[Parm  0<parm 0 L>] ?+0(GPR15)
 [     0x2aa13ab47e0]	                       LG      &GPR_0019, Parm[Parm  1<parm 1 L>] ?+8(GPR15)
 [     0x2aa13ab4900]	                       XGR     GPR_0021,GPR_0021	
 [     0x2aa13ab49b0]	                       CLR     &GPR_0019,&GPR_0017	
 [     0x2aa13ab4a60]	                       BRC     BH(0x8), Label L0018	
 [     0x2aa13ab5db0]	                       assocreg
 [     0x2aa13ab5810]	                       Label L0020:	# (Start of internal control flow)	
 POST:
 {AssignAny:&GPR_0019:R} {AssignAny:&GPR_0017:R} {AssignAny:GPR_0021:R}
 [     0x2aa13ab4ca0]	                       CLC     #389 0(200,&GPR_0019),#388 0(&GPR_0017)
 [     0x2aa13ab4d60]	                       Label L0016:	
 [     0x2aa13ab4e20]	                       IPM     GPR_0021,GPR_0021	
 [     0x2aa13ab4ed0]	                       SLLG    GPR_0021,GPR_0021,34
 [     0x2aa13ab4f90]	                       SRAG    GPR_0021,GPR_0021,62
 [     0x2aa13ab56a0]	                       assocreg
 [     0x2aa13ab5e60]	                       Label L0021:	# (End of internal control flow)	
 [     0x2aa13ab5050]	                       Label L0018:	
 POST:
 {AssignAny:&GPR_0019:R} {AssignAny:&GPR_0017:R} {AssignAny:GPR_0021:R}
 [     0x2aa13ab5fd0]	                       LGFR    GPR_0021,GPR_0021	
 [     0x2aa13ab6630]	                       assocreg
 [     0x2aa13ab6080]	                       retn    	
 POST:
 {GPR2:GPR_0021:R}
 [     0x2aa13ab6d90]	                       fence   Relative [ 0x2aa13a269b4 ] BBEnd </block_2>	
 [     0x2aa13ab7340]	                       assocreg
```
The instructions looks like this for variable load:
```
 [     0x2aa3b55b210]	                       proc    	
 [     0x2aa3b55b310]	                       fence   Relative [ 0x2aa3b4eba90 ] BBStart <block_2> (frequency 10000)	
 [     0x2aa3b55b980]	                       L       GPR_0019, Parm[Parm  2<parm 2 I>] ?+16(GPR15)
 [     0x2aa3b5798c0]	                       LG      &GPR_0020, Parm[Parm  0<parm 0 L>] ?+0(GPR15)
 [     0x2aa3b579b10]	                       LG      &GPR_0022, Parm[Parm  1<parm 1 L>] ?+8(GPR15)
 [     0x2aa3b579e30]	                       XGR     GPR_0024,GPR_0024	
 [     0x2aa3b579ee0]	                       CLR     &GPR_0022,&GPR_0020	
 [     0x2aa3b579f90]	                       BRC     BH(0x8), Label L0018	
 [     0x2aa3b57c1a0]	                       assocreg
 [     0x2aa3b57bc00]	                       Label L0023:	# (Start of internal control flow)	
 POST:
 {GPR14:GPR_0025:R} {GPR1:&GPR_0022:R} {GPR2:&GPR_0020:R} {GPR0:GPR_0026:R} {AssignAny:GPR_0024:R} {AssignAny:GPR_0019:R}
 [     0x2aa3b57a250]	                       LLILF   GPR_0025,0	
 [     0x2aa3b57a3e0]	                       SRAG    GPR_0026,GPR_0019,8
 [     0x2aa3b57a4a0]	                       BRC     BH(0x8), Label L0021	
 [     0x2aa3b57a560]	                       Label L0020:	
 [     0x2aa3b57a7a0]	                       CLC     #390 0(256,&GPR_0022),#389 0(&GPR_0020)
 [     0x2aa3b57a860]	                       BRC     BNH(0x6), Label L0016	
 [     0x2aa3b57a9e0]	                       LA      &GPR_0020,#391 256(&GPR_0020)
 [     0x2aa3b57ab50]	                       LA      &GPR_0022,#392 256(&GPR_0022)
 [     0x2aa3b57ac00]	                       BRCT    GPR_0026,Label L0020	
 [     0x2aa3b57acc0]	                       Label L0021:	
 [     0x2aa3b57ade0]	                       RISBGN  GPR_0019,GPR_0019,52,187,4	
 [     0x2aa3b57af70]	                       BAS     GPR_0025,#393 0(GPR_0019,GPR_0025)
 [     0x2aa3b57b020]	                       Label L0022:	
 [     0x2aa3b57b0e0]	                       Label L0016:	
 [     0x2aa3b57b1a0]	                       IPM     GPR_0024,GPR_0024	
 [     0x2aa3b57b250]	                       SLLG    GPR_0024,GPR_0024,34
 [     0x2aa3b57b310]	                       SRAG    GPR_0024,GPR_0024,62
 [     0x2aa3b57ba90]	                       assocreg
 [     0x2aa3b57c250]	                       Label L0024:	# (End of internal control flow)	
 [     0x2aa3b57b3d0]	                       Label L0018:	
 POST:
 {GPR14:GPR_0025:R} {GPR1:&GPR_0022:R} {GPR2:&GPR_0020:R} {GPR0:GPR_0026:R} {AssignAny:GPR_0024:R} {AssignAny:GPR_0019:R}
 [     0x2aa3b57c3c0]	                       LGFR    GPR_0024,GPR_0024	
 [     0x2aa3b57ca20]	                       assocreg
 [     0x2aa3b57c470]	                       retn    	
 POST:
 {GPR2:GPR_0024:R}
 [     0x2aa3b59a130]	                       fence   Relative [ 0x2aa3b4eba94 ] BBEnd </block_2>	
 [     0x2aa3b59a6e0]	                       assocreg
```

ehsankianifar commented:
Finally, the arraycmpSIMDHelper implements the vectorized cmp version of instructions and it is shared between both arraycmp and arraycmplen and other evaluators. like arraycmpHelper, it accepts a target label and target condition and could either return a result (len or -1/0/1 or 1/0/2) or jump to the target if the condition is satisfied!
like arrycmpHelper, I set the result if it is requested and compare the address registers. jumps either to the end or to the target is the addresses are equal.
the instructions look like this after my changes:
```
 [     0x2aa2fd01130]	                       proc    	
 [     0x2aa2fd01230]	                       fence   Relative [ 0x2aa2fc919b0 ] BBStart <block_2> (frequency 10000)	
 [     0x2aa2fd01830]	                       LG      &GPR_0017, Parm[Parm  1<parm 1 L>] ?+8(GPR15)
 [     0x2aa2fd1f7e0]	                       LG      &GPR_0019, Parm[Parm  0<parm 0 L>] ?+0(GPR15)
 [     0x2aa2fd1faa0]	                       LG      GPR_0023, Parm[Parm  2<parm 2 J>] ?+16(GPR15)
 [     0x2aa2fd1fe50]	                       LGR     GPR_0027,GPR_0023	
 [     0x2aa2fd20020]	                       CLR     &GPR_0017,&GPR_0019	
 [     0x2aa2fd200d0]	                       BRC     BH(0x8), Label L0017	
 [     0x2aa2fd20190]	                       SLGFI   GPR_0023,1	
 [     0x2aa2fd20270]	                       Label L0016:	# (Start of internal control flow)	
 [     0x2aa2fd203f0]	                       VLL     VRF_0024,GPR_0023,#389 0(&GPR_0017)
 [     0x2aa2fd20570]	                       VLL     VRF_0025,GPR_0023,#390 0(&GPR_0019)
 [     0x2aa2fd20630]	                       VFENEBS VRF_0026,VRF_0024,VRF_0025
 [     0x2aa2fd206f0]	                       BRC     BNH(0x6), Label L0018	
 [     0x2aa2fd20870]	                       LA      &GPR_0017,#391 16(&GPR_0017)
 [     0x2aa2fd209e0]	                       LA      &GPR_0019,#392 16(&GPR_0019)
 [     0x2aa2fd20a90]	                       SLGFI   GPR_0023,16	
 [     0x2aa2fd20b70]	                       BRC     MASK4(0x3), Label L0016	
 [     0x2aa2fd20c30]	                       BRC     NOP(0xf), Label L0017	
 [     0x2aa2fd20cf0]	                       Label L0018:	
 [     0x2aa2fd20db0]	                       SLGFI   GPR_0027,1	
 [     0x2aa2fd20e90]	                       SGR     GPR_0027,GPR_0023	
 [     0x2aa2fd21000]	                       VLGVB   GPR_0023,VRF_0026,#393 7
 [     0x2aa2fd210c0]	                       AGR     GPR_0027,GPR_0023	
 [     0x2aa2fd21830]	                       assocreg
 [     0x2aa2fd21290]	                       Label L0017:	# (End of internal control flow)	
 POST:
 {AssignAny:&GPR_0017:R} {AssignAny:&GPR_0019:R} {AssignAny:GPR_0023:R} {AssignAny:VRF_0024:R} {AssignAny:VRF_0025:R} {AssignAny:VRF_0026:R} {AssignAny:GPR_0027:R}
 [     0x2aa2fd21f40]	                       assocreg
 [     0x2aa2fd21990]	                       retn    	
 POST:
 {GPR2:GPR_0027:R}
 [     0x2aa2fd22910]	                       fence   Relative [ 0x2aa2fc919b4 ] BBEnd </block_2>	
 [     0x2aa2fd22ec0]	                       assocreg
```

ehsankianifar commented:
**Test and confirmation:**
I used the existing compiler tril tests implemented in ArrayTest class to test my changes. in many cases the existing tests could not reach the code path that I required so I needed to tweak the test or the code to reach that code path. To confirm that it actually branches of the main stream if the addresses are equal, I used gdb and executed the instructions line by line.

r30shah commented on compiler/z/codegen/OMRTreeEvaluator.cpp:
```diff
@@ -10513,444 +10510,350 @@ OMR::Z::TreeEvaluator::arraycmpHelper(TR::Node *node,
-         baseSource2Ref = generateS390MemoryReference(cg->evaluate(source2Node), 0, cg);
-         cg->decReferenceCount(source1Node);
-         cg->decReferenceCount(source2Node);
+         generateS390CompareAndBranchInstruction(cg, TR::InstOpCode::CLR, node, tempSource1Reg, tempSource2Reg, TR::InstOpCode::COND_BE, shortCircuitLabel);
```
Given that the only `targetLabel` is different in this `if-else` can we use the ternary?

r30shah commented on compiler/z/codegen/OMRTreeEvaluator.cpp:
```diff
@@ -11885,18 +11896,20 @@ OMR::Z::TreeEvaluator::arraycmplenEvaluator(TR::Node * node, TR::CodeGenerator *
    TR::RegisterDependencyConditions * dependencies = cg->createDepsForRRMemoryInstructions(node, firstPair, secondPair);
 
+   generateRRInstruction(cg, TR::InstOpCode::getLoadRegOpCode(), node, resultReg, orgLen);
+   generateS390CompareAndBranchInstruction(cg, TR::InstOpCode::CLR, node, firstBaseReg, secondBaseReg, TR::InstOpCode::COND_BE, shortCircuitLabel);
```
Similar comment as above, baseRegs would be 64-Bit on 64-Bit platform and would be 32-bit on 32-bit platform. 

You can use the query https://github.com/eclipse/omr/blob/83cb59850cf7c9da28183aeeaa7747908160835a/compiler/z/codegen/InstOpCode.cpp#L458 to automatically get the instOpCode that is appropriate for the platform.

r30shah commented on compiler/z/codegen/OpMemToMem.cpp:
```diff
@@ -1636,6 +1635,8 @@ MemCmpVarLenSignMacroOp::generate(TR::Register* dstReg, TR::Register* srcReg, TR
    {
    TR::Instruction *cursorBefore = cursor;
    _resultReg = _cg->allocateRegister();
+   generateRRInstruction(_cg, TR::InstOpCode::getXORRegOpCode(), _rootNode, _resultReg, _resultReg);
```
Why can't we use the subtract operation here and branch if result is zero ? This will save one extra instruction.

r30shah reviewed (changes requested):
@ehsankianifar for some reason, I have left some initial thoughts on the changes you made. Will give a detailed review soon and inspect the Instruction selection logs you have posted here.

ehsankianifar commented on compiler/z/codegen/OMRTreeEvaluator.cpp:
```diff
@@ -10513,16 +10510,73 @@ OMR::Z::TreeEvaluator::arraycmpHelper(TR::Node *node,
+            cg,
+            TR::InstOpCode::getCmpLogicalRegOpCode(),
+            node,
+            cg->evaluate(source1Node),
```
we evaluate here as well as in the creation of mem ref. Maybe makes more sense assign it to a named register first. I can Also move this code after mem ref creation and compare base registers of mem refs!

r30shah commented on compiler/z/codegen/OMRTreeEvaluator.cpp:
```diff
@@ -10513,16 +10510,73 @@ OMR::Z::TreeEvaluator::arraycmpHelper(TR::Node *node,
+            {
+            // since forSS was true in the generation of the memory references, there will not be any index registers.
+            // if there are base registers, we can compare those and conclude if are equal.
+            if(baseSource1Ref->getBaseRegister() && baseSource2Ref->getBaseRegister())
```
Yes, there should be base register (We have D(B) format or D(X,B) format memory reference). So there should always be. 

Though, can you elaborate the change here bit more?

ehsankianifar replied:
I thought it is possible to have only displacement in ss structure but I was wrong. I think it is safe to assume the index register is null here since the `forSS` flag is set. first I check to make sure the displacements are equal and if it is, I compare base registers. But It seems that when we create a memRef from node, the `_baseRegister` property is null.

ehsankianifar commented on compiler/z/codegen/OMRTreeEvaluator.cpp:
```diff
@@ -10529,6 +10527,19 @@ OMR::Z::TreeEvaluator::arraycmpHelper(TR::Node *node,
          cg->decReferenceCount(source2Node);
          }
 
+      if (baseSource1Ref->getOffset() == baseSource2Ref->getOffset())
```
Not sure if offsets are expected but later in original code, we check those therefore I decided to check as well.

r30shah commented:
@ehsankianifar When this PR is ready for review, can you squash all the commits? I will restart review process.

ehsankianifar commented:
Thanks @r30shah. I squashed my commits and did some testing. Please take a look when you have time.

r30shah commented on compiler/z/codegen/OMRTreeEvaluator.cpp:
```diff
@@ -10529,6 +10527,19 @@ OMR::Z::TreeEvaluator::arraycmpHelper(TR::Node *node,
          cg->decReferenceCount(source2Node);
          }
 
+      if (baseSource1Ref->getOffset() == baseSource2Ref->getOffset())
```
Hmm, I do see what you are trying to do here. Wouldn't this miss, a case where we have different offset and base register values, but both memory reference would point to same location ?

ehsankianifar replied:
I can load address to register to cover that case but I am not sure if it worth it since the probability of that scenario doesn't seem to be high. I am still trying to figure out how the created memRefs could have non zero offset. We create a MemRef from node which I think it just evaluate the node and set the base register (the provided displacement is zero)

r30shah replied:
I am talking about the case where we do have the symbol reference on the node. In that case, you are comparing the offset and if the offset at compile time is same, u generate the runtime test to check if the value in base register is same or not. 
For example we have a case where offset1 + base1 is equal to offset2 + base2 but offset1 != offset2 and same with base register. 

If you meant to have a quick optimize test specific to case where at compile time you know offset are same and at run time you check in that case if the base values are same or not then this branch is justified, but this will simply covering a single case.

We should look into the series of instruction generated here to know if we can inspect the memory reference pointing to the same location or not. 

Can you share the generated instructions here?

r30shah replied:
I am aware of this way of comparing, I do want to take a look at the instruction selection log to see if there is a better way.

r30shah commented on compiler/z/codegen/OMRTreeEvaluator.cpp:
```diff
@@ -11885,18 +11898,20 @@ OMR::Z::TreeEvaluator::arraycmplenEvaluator(TR::Node * node, TR::CodeGenerator *
 
    TR::RegisterDependencyConditions * dependencies = cg->createDepsForRRMemoryInstructions(node, firstPair, secondPair);
 
+   generateRRInstruction(cg, TR::InstOpCode::getLoadRegOpCode(), node, resultReg, orgLen);
```
Do you have a selection log that executes this path? If not can you explain how moving loading the origLen before CLCL different compared to after executing CLCL?

ehsankianifar replied:
In case that addresses are equal and we need to short circuit, the result should be equal to the length since this is `arraycmplen()` evaluator and we return the address of the first mismatch!

r30shah commented on compiler/z/codegen/OMRTreeEvaluator.cpp:
```diff
@@ -11905,12 +11920,12 @@ OMR::Z::TreeEvaluator::arraycmplenEvaluator(TR::Node * node, TR::CodeGenerator *
    cg->decReferenceCount(secondBaseAddr);
    cursor->setDependencyConditions(dependencies);
 
+   generateS390LabelInstruction(cg, TR::InstOpCode::label, node, shortCircuitLabel);
```
Don't we have an internal control flow here that ends at shortCircuitLabel here?

ehsankianifar replied:
I thought the control flow is when there is a loop involved which is not the case in this method. Was I wrong?

r30shah replied:
No, that is not true. Here you have branch in your code that jumps to done label and unsuccessful branch continues execution. Local RA without properly defined internal control flow may spill register in one path but not other.

r30shah commented on compiler/z/codegen/OMRTreeEvaluator.cpp:
```diff
@@ -15783,15 +15798,31 @@ OMR::Z::TreeEvaluator::arraycmpSIMDHelper(TR::Node *node,
-   generateRILInstruction(cg, TR::InstOpCode::getSubtractLogicalImmOpCode(), node, lastByteIndexReg, 1);
-   if(needResultReg && isArrayCmpLen)
-      generateRRInstruction(cg, TR::InstOpCode::getLoadRegOpCode(), node, resultReg, lastByteIndexReg);
+   if(needResultReg)
```
Do you have a selection log executing this path? I would like to verify the dependancies attached to various branch / label

r30shah reviewed (commented):
Overall I think changes looks OK to me, but would appreciate if you can address some of the question I posted.

r30shah commented:
@ehsankianifar  Can you paste the instruction selection log?

r30shah commented on compiler/z/codegen/OMRTreeEvaluator.cpp:
```diff
@@ -5025,13 +5025,13 @@ generateS390CompareAndBranchOpsHelper(TR::Node * node, TR::CodeGenerator * cg, T
 
                TR::TreeEvaluator::arraycmpHelper(nonConstNode,
                      cg,
-                     false,      //isWideChar
```
Just wondering, why the comment was removed for `isWideChar` and `return102` ?

r30shah commented on compiler/z/codegen/OMRTreeEvaluator.cpp:
```diff
@@ -10438,11 +10436,11 @@ OMR::Z::TreeEvaluator::arraycmpHelper(TR::Node *node,
       {
       // An empirical study has showed that CLC is faster for all array sizes if the number of bytes to copy is known to be constant
       if (!node->getChild(2)->getOpCode().isLoadConst())
-          return TR::TreeEvaluator::arraycmpSIMDHelper(node, cg, compareTarget, ificmpNode, needResultReg, return102);
```
Can you remind why `return102` is set to `true` by-default here?

r30shah commented on compiler/z/codegen/OMRTreeEvaluator.cpp:
```diff
@@ -10438,11 +10436,11 @@ OMR::Z::TreeEvaluator::arraycmpHelper(TR::Node *node,
    {
    TR::Compilation *comp = cg->comp();
 
-   if (TR::isJ9() && !comp->getOption(TR_DisableSIMDArrayCompare) && cg->getSupportsVectorRegisters() && !isWideChar)
```
Checking out the code for `isWideChar`, I see that we double the length for the operand, which may suggests that the code was added to handle comparison of Strings that are UTF-16.   @joransiu  Do you remember why we had this `isWideChar` code ? Checking the current code, it is set to false always and I do not see other codegen doing anything special for this. Want to make sure it is safe to remove it. 

Tagging @jdmpapin / @dchopra001  here as well , if they have any more context behind this.

r30shah reviewed (changes requested):
@ehsankianifar  thanks for logically dividing these changes, I have asked couple of question to you and others to clarify some doubts. Will give final reviews based on the answers.

ehsankianifar commented:
fixed in:
https://github.com/eclipse/omr/pull/7231
https://github.com/eclipse/omr/pull/7229
https://github.com/eclipse/omr/pull/7228
https://github.com/eclipse/omr/pull/7222
https://github.com/eclipse/omr/pull/7220

ehsankianifar commented:
> @ehsankianifar Is this PR still relevant ?

@r30shah no, I close it. Thanks for checking.
